### ISSUE TITLE
TLS overhead x4

### ISSUE THREAD

**Issue by gr8den** [restic/rest-server#205]
Hello!
When I enable `--tls` option in rest-server speed of backup become slower ~4 times (1MB/s vs 4MB/s)

Other options: `--append-only`, .htpasswd auth used, self-signed cert
Ping between restic and rest-server machines ~90ms

Setting `-o rest.connections=` to 2 or 10 change almost nothing when it is used with --tls option
But using `-o rest.connections=10` & --tls=0 increase performance to 11MB/s (maximum bandwidth)

Rest-server version: https://github.com/restic/rest-server/commit/a8cd3f218de825125ad845a64244a68e402d3c0c
Restic version: `restic 0.14.0 compiled with go1.19 on darwin/amd64`

**Comment by MichaelEischer** [comment]
I don't suppose the server is anyhow CPU-bound?

Does the performance also drop if both restic and the rest-server run on the same host? Does setting
```
GODEBUG=http2client=0  # disable HTTP/2 client support
GODEBUG=http2server=0  # disable HTTP/2 server support
```
on the client/server side help?

**Comment by deajan** [comment]
Maybe worth mentionning, I let Haproxy offload the TLS encryption, it's probably way more optimized since it's it's primary job.
You should give that a go.
On the plus side, haproxy directly implements with letsencrypt.

**Comment by MichaelEischer** [comment]
Not enough information to reproduce.

**Comment by eliyen** [comment]
I also ran into this issue.

_restic 0.18.1 compiled with go1.25.1 on windows/amd64
rest-server 0.14.0 compiled with go1.24.3 on linux/amd64_

Running rest-server over HTTP, I'm able to fully saturate a 500 Mbit/s downlink when restoring a snapshot. However, running rest-server over HTTPS (`--tls` `--tls-key` `--tls-cert` enabled), I only get around 120 Mbit/s when restoring. In both cases, I set `rest.connections=32`

Setting `GODEBUG=http2client=0` on the client didn't seem to have any effect. However, setting `GODEBUG=http2server=0` on the server, I was able to get full download speeds even when running over HTTPS. I'm guessing there's an issue related to Go's HTTP2 server implementation. Should HTTP2 be disabled by default in rest-server until this is fixed?
